### ISSUE TITLE
Optional<T[]> and Collection<T[]> Produce Incorrect Immutables

### ISSUE THREAD

**Issue by tpetracca** [palantir/atlasdb#2020]
In debugging why error-prone was failing in #2016 I stumbled across the fact that Immutables do not support having Array parameters for Optionals or Collections. error-prone catches this for java's Optional in your immutable interface, but does not for guava's Optional because of the different ways Immutables handles generating the code for each case. See https://github.com/palantir/atlasdb/pull/2016#issuecomment-307909022 for an in-depth example.

Basically this means in our current code we have a bunch of Immutable code who equals and hashCode methods are bad.

**Comment by tpetracca** [comment]
So it turns out that it's even edge-casier. The specific java Optional example in 2016 only shows because of the way immutables generates the specific way we're using the Optional in that class. Another example in `CheckAndSetRequest` which uses
```
    @Value.Default
    public Optional<byte[]> oldValue() {
        return Optional.empty();
    }
```
is also wrong in the Immutable code
```
private final Optional<byte[]> oldValue;
```
```
  private boolean equalTo(ImmutableCheckAndSetRequest another) {
    return table.equals(another.table)
        && cell.equals(another.cell)
        && oldValue.equals(another.oldValue)
        && Arrays.equals(newValue, another.newValue);
  }
```
and doesn't catch it because it's using `Optional<T[]>` and not `@Nullable T[]` like it was in `SweepResults`.

Going to take a pass at all the places where we have some form of `Optional<T[]>` or `Collection<T[]>`

**Comment by schlosna** [comment]
We should really use a `ByteString` (e.g. from `protobuf` or `okio`), and ideally interact with it via reading a `ByteBuffer`. Might be worth pulling a shaded, independent version that can be utilized in APIs.

**Comment by tpetracca** [comment]
Primary offenders on the `Optional` side of the issue.
```
$ git grep "Optional<.*\[\]>"
atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/CheckAndSetRequest.java:    public Optional<byte[]> oldValue() {
atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/SweepResults.java:    public Optional<byte[]> getPreviousStartRow() {
atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/SweepResults.java:    public Optional<byte[]> getNextStartRow() {
atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java:        java.util.Optional<Entry<TableReference, byte[]>> match =
atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java:        java.util.Optional<byte[]> oldValue = request.oldValue();
atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/SweepCommand.java:            Optional<byte[]> startRow = Optional.of(entry.getValue());
atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/SweepCommand.java:    private String encodeStartRow(Optional<byte[]> rowBytes) {
atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/SweepCommand.java:    private String encodeEndRow(Optional<byte[]> rowBytes) {
atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java:        Optional<byte[]> oldValue = request.oldValue();
```
All of which come down to two main offenders.
`SweepResults`
```
public abstract Optional<byte[]> getPreviousStartRow();

public abstract Optional<byte[]> getNextStartRow();
```
and `CheckAndSetRequest`
```
@Value.Default
public Optional<byte[]> oldValue() {
    return Optional.empty();
}
```

I think in both cases we can change to protobuf ByteString as you suggest without breaking wire-format/back-compat. Going to take a stab at this.

**Comment by tpetracca** [comment]
Actually to not break wire-format, we'd have to make the `Optional<ByteString>` derived from a `@Nullable byte[]` at which point we should just use the byte[] directly.

**Comment by tpetracca** [comment]
@gbonik @gsheasby  do you know if we ever actually store a `SweepResults` blob anywhere? If we're not then it should be safe to break the wire-format of right? In which case I can use protobuf's ByteString? Going the @Nullable route is possible here, but unideal for a couple of reasons:
1. Immutables will re-copy the byte[] (if it exists) on every call to the method fetching it. We can be smart with the code to avoid re-do'ing this a lot if needed.
2. Generally this stuff is just referenced a lot so the diff for the Nullable version was larger than I would have liked it to be. Although it's unclear that this gets better if we switch it to ByteString because we still have to convert it back to byte[] in a bunch of places.

**Comment by gbonik** [comment]
@tpetracca I don't think we ever persist SweepResults as is. Also not sure why it's in the atlasdb-api project, since it's an intermediate result object that is more like an implementation detail. So changing the type should be safe.

Regarding the ByteString alternatives, I'd probably vote for a very transparent wrapper class that also wouldn't copy the array on creation, as well as in its "toByteArray()" method. We treat the byte arrays as immutable objects anyway; see the Cell class for example.

**Comment by gsheasby** [comment]
I would be happy see SweepResults moved out of atlasdb-api if possible, assuming that it's never sent over the wire anywhere.

**Comment by hsaraogi** [comment]
Check if Optional<T[]>  or Collection<T[]> is used elsewhere and move the usages to `Optional<List<T>>` or `Collection<List<T>>`. We will be happy to accept a PR too.

**Comment by tpetracca** [comment]
Also where performance is important we shouldn't use arrays at all as they need to be fully copied due to the mutable nature of them.
https://immutables.github.io/immutable.html#array-collection-and-map-attributes

**Comment by gsheasby** [comment]
Backlogging - we aren't going to get to this ourselves soon.

**Comment by felixdesouza** [comment]
I believe I inadvertently fixed this, I went with a new field that is:

    @Nullable @JsonProperty("oldValue") abstract byte[] oldValueNullable

This does the correct behaviour r.e. arrays and equals.

For code back-compat:
    
    @JsonIgnore
    public Optional<byte[]> oldValue() { ... }
